deeppavlov/dataset_iterators/multitask_pal_bert_iterator.py

```diff
@@ -0,0 +1,322 @@
+                if task not in united:
+                    united[task] = data
+                else:
+                    united[task] = united[task] + data
+        return united
+
+    def _init_sample_tasks_batch(self, batch_size=1):
+        iters = [iter_ for iter_ in self.task_iterators.values()]
+        sample_batch = [i.gen_batches(batch_size).__next__() for i in iters]
+        self.sample_x_instances = []
+        self.sample_y_instances = []
+        for sample_task_batch in sample_batch:
+            self.sample_x_instances.append(sample_task_batch[0])
+            self.sample_y_instances.append(sample_task_batch[1])
+
+    def gen_batches(
+        self, batch_size: int, data_type: str = "train", shuffle: bool = None
+    ) -> Iterator[Tuple[tuple, tuple]]:
+        """Generate batches and expected output to train neural networks. Batches from
+        task iterators are united into one batch. Every element of the largest dataset
+        is used once whereas smaller datasets are repeated until their size is equal to
+        the largest dataset.
+
+        Args:
+            batch_size: number of samples in batch
+            data_type: can be either 'train', 'test', or 'valid'
+            shuffle: whether to shuffle dataset before batching
+
+        Yields:
+            A tuple of a batch of inputs and a batch of expected outputs.
+            Inputs and outputs are tuples. Element of inputs or outputs is a tuple which
+            elements are x values of merged tasks in the ordertasks are present in
+            `tasks` argument of `__init__` method.
+        """
+        max_task_data_len = self.max_task_data_len[data_type]
+        size_of_last_batch = max_task_data_len % batch_size
+        if size_of_last_batch == 0:
+            size_of_last_batch = batch_size
+
+        n_batches = math.ceil(max_task_data_len / batch_size)
+
+        if data_type == "train":
+            generators = [
+                RepeatBatchGenerator(iter_, batch_size, data_type, shuffle)
+                for iter_ in self.task_iterators.values()
+            ]
+
+            # one sample batch of batch_size 1 for each task
+            if not self.sample_x_instances or not self.sample_y_instances:
+                self._init_sample_tasks_batch(batch_size)
+
+            # probs only required while training
+            probs = self._get_probs("train")
+            for step in range(self.steps_per_epoch):
+                if (
+                    self.steps_taken + 1
+                ) % self.gradient_accumulation_steps or self.task_id:
+                    self.task_id = np.random.choice(self.n_tasks, p=probs)
+                batch = generators[self.task_id].__next__()
+                x_instances = self.sample_x_instances.copy()
+                y_instances = self.sample_y_instances.copy()
+                x_instances[self.task_id] = batch[0]
+                y_instances[self.task_id] = batch[1]
+                b = (
```

Comment:
it's better to name b, as well as it's components so as to be clear what it means

Reply:
changed